### PR TITLE
Persistent firmware UUID

### PR DESCRIPTION
**Pull request by dankenigsberg**
### What this PR does
Before this PR:
KubeVirt regenerated Firmware.UUID for VMI whenever the VM is started.

After this PR:
KubeVirt generated the Firmware.UUID once and persists in vm.Spec. Eventually, all VMs would have Firmware.UUID persisted, and we would be free to fix https://github.com/kubevirt/kubevirt/issues/13156

Fixes #

### Why we need it and why it was done in this way
We need to generate a Firmware.UUID that is universally unique AND keeps the identity of existing VMs.
This PR is one step for the fix of https://github.com/kubevirt/kubevirt/issues/13156 . Another fix would follow once we are certain that all supported KubeVirt VMs have Firmware.UUID persisted in their spec.

### Special notes for your reviewer

We can expedite the fix for new VMs by adding a mutating webhook that generates a good UUID for new VMs: https://github.com/kubevirt/kubevirt/pull/12085. I prefer not to do this right now because webhooks are slow and fragile, a liability that we'd rather avoid.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
```release-note
If Firmware.UUID is missing in vm.Spec, it is generated and persisted to ensure that the identification of existing VMs would not change once we fix UUID generation logic.
```

### CODE REVIEW

**Comment by Acedus** [comment]
Quick heads upon this: https://github.com/kubevirt/kubevirt/pull/11199, https://github.com/kubevirt/kubevirt/pull/12085.

I'm aware this is a draft but current logic states that firmware UUID will be regenerated according to the new namespace+name convention (unless firmware UUID in VMs is set, which AFAIK isn't the case by default).

This can be problematic due to things like: https://github.com/kubevirt/kubevirt/blob/37b59b2e0f55ae845d6fe6966dfec24092acd9b1/pkg/cloud-init/cloud-init.go#L117-L122

Where upon restart with the new logic cloud-init would rerun due to an `instance-id` change.

It might be worth considering an approach to initially sync the firmware UUID of VMIs with their VMs, so for at least one version, all VMs in a cluster would have a stable firmware UUID (unless explicitly specified by the user). This way, the new logic would only apply to newly created VMs, though we’d still need to consider potential risks with VM backups.

FYI @fabiand

File: pkg/virt-controller/watch/vm/vm.go [hunk 1]
```diff
@@ -2056,28 +2054,27 @@ func hasStopRequestForVMI(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineI
...
 		return
 	}
 
-	vmi.Spec.Domain.Firmware.UUID = types.UID(uuid.NewSHA1(firmwareUUIDns, []byte(vmi.ObjectMeta.Name)).String())
+	// This code generates a non-uniqe identification. Two VMs with the
+	// same name but on different namespaces or on different clusters get
+	// the same ID. We have to keep this for a while because modifying this
+	// logic would modify the ID of existing VMs.
+	// https://github.com/kubevirt/kubevirt/issues/13156
+	vm.Spec.Template.Spec.Domain.Firmware.UUID = types.UID(uuid.NewSHA1(firmwareUUIDns, []byte(vmi.ObjectMeta.Name)).String())
```
Comment (iholder101) [inline pkg/virt-controller/watch/vm/vm.go:2077]: This part is problematic, as it's invalid to modify the vm here as it's the cached object from the informer's store, which will leave the object in an inconsistent state.

This code is eventually being called from here:
https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/pkg/virt-controller/watch/vm/vm.go#L3280

a few lines later, the vm is being copied. Only the copy should be modified, leaving the cached object unchanged:
https://github.com/kubevirt/kubevirt/blob/96c9f1180eb8770d58a1f74eb70f0a569d074e8e/pkg/virt-controller/watch/vm/vm.go#L3294

So even if we follow this approach, we'd have to do it in the scope above. Let me hold the PR in the meantime.

/hold

**Review by iholder101 — COMMENTED** [review]
Thank you @dankenigsberg! This is indeed a recurring problem that needs to be fixed.

An alternative to the approach in this PR is adding a firmware UUID field to the VM's status. The obvious con is that we'll be adding an API field which is probably not very important outside the scope of this bug.

However, the pro is that this approach uses the right separation between spec (the user's desired state) and status (the actual state). Ideally, the spec should contain only whatever the user cares about and leave everything else blank.

File: pkg/virt-controller/watch/vm/vm.go (hunk 1 above)
Comment (dankenigsberg) [reply pkg/virt-controller/watch/vm/vm.go:2077]: > An alternative to the approach in this PR is adding a firmware UUID field to the VM's status. The obvious con is that we'll be adding an API field which is probably not very important outside the scope of this bug.

I trust you that my implementation is bad, but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.

Can you suggest another place where one can fill Firmware.UUID of existing VMs, e.g during virt-controller startup?

Comment (iholder101) [reply pkg/virt-controller/watch/vm/vm.go:2077]: > but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.

Sorry, perhaps I'm missing something, but why wouldn't a status field solve this problem?
My idea is that whenever the VM would boot for the first time, and create a UUID, this UUID will be saved to the VM's status in a persistent manner. This way whenever the VM would boot, we would first try to see if the status UUID field is populated, and if it is we'd use this saved UUID.

Am I missing anything?

Comment (Acedus) [reply pkg/virt-controller/watch/vm/vm.go:2077]: I’d like to note that this is around the fourth time the topic of persistent firmware UUID is discussed. Each conversation has its own merits, but I particularly agree with  https://github.com/kubevirt/kubevirt/pull/12085#issuecomment-2275384536. This issue calls for a design proposal IMO.

Comment (vladikr) [reply pkg/virt-controller/watch/vm/vm.go:2077]: > > but 'status' is unlikely to help. VMs should have a persistent firmware UUID to be reused on the next time the VM is started.
> 
> Sorry, perhaps I'm missing something, but why wouldn't a status field solve this problem? My idea is that whenever the VM would boot for the first time, and create a UUID, this UUID will be saved to the VM's status in a persistent manner. This way whenever the VM would boot, we would first try to see if the status UUID field is populated, and if it is we'd use this saved UUID.
> 
> Am I missing anything?

I like this idea. It fits the same pattern we have with CurrentCPUTopology and CurentMemeory, as well as the pattern we were trying to push for the migrated volumes

Comment (vladikr) [reply pkg/virt-controller/watch/vm/vm.go:2077]: > However, the pro is that this approach uses the right separation between spec (the user's desired state) and status (the actual state). Ideally, the spec should contain only whatever the user cares about and leave everything else blank.

Patching the spec of a running VM can be tricky. When the `rolloutStrategy` is set to `liveUpdate`, patching the VM spec would mean that we will raise the restart required condition.
Most of the VM should be running after KubeVirt's upgrade, hence we will run into this issue, but not if we will store the UUID in the VM status.

Comment (iholder101) [reply pkg/virt-controller/watch/vm/vm.go:2077]: Thank you all for the discussion.
Let's please continue it on @dasionov's design proposal here: https://github.com/kubevirt/community/pull/347.

I've tried to list all of the different alternatives in https://github.com/kubevirt/community/pull/347#pullrequestreview-2420156264. Let me know if you have more alternatives in mind.

**Comment by dankenigsberg** [comment]
Closing to let https://github.com/kubevirt/community/pull/347 conclude

**Pull request closed without merging**